{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2283198e-8490-41e0-9488-9f2dc11573fe",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-07-03T12:37:31.656404Z",
     "iopub.status.busy": "2024-07-03T12:37:31.655907Z",
     "iopub.status.idle": "2024-07-03T12:37:31.659081Z",
     "shell.execute_reply": "2024-07-03T12:37:31.658571Z",
     "shell.execute_reply.started": "2024-07-03T12:37:31.656376Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# !pip install scglue==0.3.2 scanpy networkx\n",
    "# !pip install  --user scikit-misc\n",
    "# !conda install -c bioconda bedtools -y\n",
    "# !pip install pyarrow cytoolz\n",
    "# !pip install pyscenic==0.12.1\n",
    "# !conda install -c pytorch -c nvidia faiss-gpu=1.8.0 -y\n",
    "# !conda install pytorch::faiss-gpu\n",
    "# !pip install numpy==1.23.4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e20aa02e-8cff-49e1-aac8-96929e95249b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-07-03T12:37:31.660452Z",
     "iopub.status.busy": "2024-07-03T12:37:31.660207Z",
     "iopub.status.idle": "2024-07-03T12:37:31.664245Z",
     "shell.execute_reply": "2024-07-03T12:37:31.663706Z",
     "shell.execute_reply.started": "2024-07-03T12:37:31.660431Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import anndata as ad\n",
    "import networkx as nx\n",
    "import scanpy as sc\n",
    "import scglue\n",
    "from matplotlib import rcParams\n",
    "import os \n",
    "import subprocess\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "work_dir = 'output/'\n",
    "data_dir = '../../../resources/multiomics'\n",
    "input_dir = 'input/'\n",
    "os.makedirs(work_dir, exist_ok=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "33484e23-99eb-4c95-821d-2a93bfc5d161",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-07-03T12:37:44.480855Z",
     "iopub.status.busy": "2024-07-03T12:37:44.480482Z",
     "iopub.status.idle": "2024-07-03T12:37:44.483812Z",
     "shell.execute_reply": "2024-07-03T12:37:44.483249Z",
     "shell.execute_reply.started": "2024-07-03T12:37:44.480829Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# pd.read_csv('output/pruned_grn.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "881d15ac-d358-42e2-98ce-7252af3cdc0d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-07-02T09:11:24.881509Z",
     "iopub.status.busy": "2024-07-02T09:11:24.881116Z",
     "iopub.status.idle": "2024-07-02T09:11:24.889230Z",
     "shell.execute_reply": "2024-07-02T09:11:24.888707Z",
     "shell.execute_reply.started": "2024-07-02T09:11:24.881478Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def preprocess(rna, atac):\n",
    "    rna.layers[\"counts\"] = rna.X.copy()\n",
    "    sc.pp.highly_variable_genes(rna, n_top_genes=2000, flavor=\"seurat_v3\")\n",
    "    sc.pp.normalize_total(rna)\n",
    "    sc.pp.log1p(rna)\n",
    "    sc.pp.scale(rna)\n",
    "    sc.tl.pca(rna, n_comps=100, svd_solver=\"auto\")\n",
    "    sc.pp.neighbors(rna, metric=\"cosine\")\n",
    "    sc.tl.umap(rna)\n",
    "    \n",
    "    scglue.data.lsi(atac, n_components=100, n_iter=15)\n",
    "    sc.pp.neighbors(atac, use_rep=\"X_lsi\", metric=\"cosine\")\n",
    "    sc.tl.umap(atac)\n",
    "    \n",
    "    scglue.data.get_gene_annotation(\n",
    "        rna, gtf=f\"{input_dir}/gencode.v45.annotation.gtf.gz\",\n",
    "        gtf_by=\"gene_name\"\n",
    "    )\n",
    "    \n",
    "    rna = rna[:, ~rna.var.chrom.isna()]\n",
    "    \n",
    "    split = atac.var_names.str.split(r\"[:-]\")\n",
    "    atac.var[\"chrom\"] = split.map(lambda x: x[0])\n",
    "    atac.var[\"chromStart\"] = split.map(lambda x: x[1]).astype(int)\n",
    "    atac.var[\"chromEnd\"] = split.map(lambda x: x[2]).astype(int)\n",
    "    \n",
    "    guidance = scglue.genomics.rna_anchored_guidance_graph(rna, atac)\n",
    "    \n",
    "    scglue.graph.check_graph(guidance, [rna, atac])\n",
    "    \n",
    "    column_names = [\n",
    "    \"chrom\",\n",
    "    \"gene_type\",\n",
    "    \"gene_id\",\n",
    "    \"hgnc_id\",\n",
    "    \"havana_gene\",\n",
    "    \"tag\",\n",
    "    \"score\",\n",
    "    \"strand\",\n",
    "    \"thickStart\",\n",
    "    \"thickEnd\",\n",
    "    \"itemRgb\",\n",
    "    \"blockCount\",\n",
    "    \"blockSizes\",\n",
    "    \"blockStarts\",\n",
    "    \"artif_dupl\",\n",
    "    \"highly_variable_rank\"\n",
    "    ]\n",
    "    rna.var[column_names] = rna.var[column_names].astype(str)\n",
    "\n",
    "    rna.write(f'{work_dir}/rna.h5ad')\n",
    "    atac.write(f'{work_dir}/atac.h5ad')\n",
    "    nx.write_graphml(guidance, f'{work_dir}/guidance.graphml.gz')\n",
    "# rna = ad.read_h5ad(f\"{data_dir}/adata_rna.h5ad\")\n",
    "# atac = ad.read_h5ad(f\"{data_dir}/adata_atac.h5ad\")\n",
    "# preprocess(rna, atac)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "76fe932a-5f6b-4328-bda6-dfe311d8a120",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-07-02T13:24:32.240165Z",
     "iopub.status.busy": "2024-07-02T13:24:32.239641Z",
     "iopub.status.idle": "2024-07-02T13:26:11.941500Z",
     "shell.execute_reply": "2024-07-02T13:26:11.940824Z",
     "shell.execute_reply.started": "2024-07-02T13:24:32.240137Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def training():\n",
    "    rna = ad.read_h5ad(f\"{work_dir}/rna.h5ad\")\n",
    "    atac = ad.read_h5ad(f\"{work_dir}/atac.h5ad\")\n",
    "    guidance = nx.read_graphml(f\"{work_dir}/guidance.graphml.gz\")\n",
    "    scglue.models.configure_dataset(\n",
    "        rna, \"NB\", use_highly_variable=False,\n",
    "        use_layer=\"counts\", use_rep=\"X_pca\"\n",
    "    )\n",
    "    scglue.models.configure_dataset(\n",
    "        atac, \"NB\", use_highly_variable=False,\n",
    "        use_rep=\"X_lsi\"\n",
    "    )\n",
    "\n",
    "    glue = scglue.models.fit_SCGLUE(\n",
    "        {\"rna\": rna, \"atac\": atac}, guidance,\n",
    "        fit_kws={\"directory\": f\"{work_dir}/glue\"}\n",
    "    )\n",
    "\n",
    "    glue.save(f\"{work_dir}/glue.dill\")\n",
    "\n",
    "    dx = scglue.models.integration_consistency(\n",
    "        glue, {\"rna\": rna, \"atac\": atac}, guidance\n",
    "    )\n",
    "\n",
    "    rna.obsm[\"X_glue\"] = glue.encode_data(\"rna\", rna)\n",
    "    atac.obsm[\"X_glue\"] = glue.encode_data(\"atac\", atac)\n",
    "    combined = ad.concat([rna, atac])\n",
    "    sc.pp.neighbors(combined, use_rep=\"X_glue\", metric=\"cosine\")\n",
    "    sc.tl.umap(combined)\n",
    "    feature_embeddings = glue.encode_graph(guidance)\n",
    "    feature_embeddings = pd.DataFrame(feature_embeddings, index=glue.vertices)\n",
    "    rna.varm[\"X_glue\"] = feature_embeddings.reindex(rna.var_names).to_numpy()\n",
    "    atac.varm[\"X_glue\"] = feature_embeddings.reindex(atac.var_names).to_numpy()\n",
    "\n",
    "    rna.write(f\"{work_dir}/rna-emb.h5ad\", compression=\"gzip\")\n",
    "    atac.write(f\"{work_dir}/atac-emb.h5ad\", compression=\"gzip\")\n",
    "    nx.write_graphml(guidance, f\"{work_dir}/guidance.graphml.gz\")\n",
    "    \n",
    "# training()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d780728e-7d43-4f4e-9a6f-5ba6717d5b21",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-07-03T12:38:17.633011Z",
     "iopub.status.busy": "2024-07-03T12:38:17.632621Z",
     "iopub.status.idle": "2024-07-03T12:38:17.635992Z",
     "shell.execute_reply": "2024-07-03T12:38:17.635421Z",
     "shell.execute_reply.started": "2024-07-03T12:38:17.632985Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# import loompy\n",
    "\n",
    "# # Open the Loom file\n",
    "# with loompy.connect(f\"{work_dir}/rna.loom\") as ds:\n",
    "#     print(ds.ra.keys())\n",
    "#     print(ds.ca.keys())\n",
    "#     print(ds.ra['name'])\n",
    "# # gene_names\n",
    "#     print(ds.ca['obs_id'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f320b129-0b71-4902-880d-68e7c6c9426a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-07-02T19:33:10.106715Z",
     "iopub.status.busy": "2024-07-02T19:33:10.106343Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "num_workers = 4\n",
    "\n",
    "def cis_inference_(num_workers):\n",
    "    ''' Infers gene2peak connections\n",
    "    '''\n",
    "    rna = ad.read_h5ad(f\"{work_dir}/rna-emb.h5ad\")\n",
    "    atac = ad.read_h5ad(f\"{work_dir}/atac-emb.h5ad\")\n",
    "    guidance = nx.read_graphml(f\"{work_dir}/guidance.graphml.gz\")\n",
    "\n",
    "    rna.var[\"name\"] = rna.var_names\n",
    "    atac.var[\"name\"] = atac.var_names\n",
    "\n",
    "    genes = rna.var.index\n",
    "    peaks = atac.var.index\n",
    "\n",
    "    features = pd.Index(np.concatenate([rna.var_names, atac.var_names]))\n",
    "    feature_embeddings = np.concatenate([rna.varm[\"X_glue\"], atac.varm[\"X_glue\"]])\n",
    "\n",
    "    skeleton = guidance.edge_subgraph(\n",
    "        e for e, attr in dict(guidance.edges).items()\n",
    "        if attr[\"type\"] == \"fwd\"\n",
    "    ).copy()\n",
    "\n",
    "    reginf = scglue.genomics.regulatory_inference(\n",
    "        features, feature_embeddings,\n",
    "        skeleton=skeleton, random_state=0\n",
    "    )\n",
    "\n",
    "\n",
    "    gene2peak = reginf.edge_subgraph(\n",
    "        e for e, attr in dict(reginf.edges).items()\n",
    "        if attr[\"qval\"] < 0.05\n",
    "    )\n",
    "\n",
    "\n",
    "    scglue.genomics.Bed(atac.var).write_bed(f\"{work_dir}/peaks.bed\", ncols=3)\n",
    "    scglue.genomics.write_links(\n",
    "        gene2peak,\n",
    "        scglue.genomics.Bed(rna.var).strand_specific_start_site(),\n",
    "        scglue.genomics.Bed(atac.var),\n",
    "        f\"{work_dir}/gene2peak.links\", keep_attrs=[\"score\"]\n",
    "    )\n",
    "\n",
    "\n",
    "\n",
    "    motif_bed = scglue.genomics.read_bed(f\"{input_dir}/JASPAR2022-hg38.bed.gz\") ## http://download.gao-lab.org/GLUE/cisreg/JASPAR2022-hg38.bed.gz\n",
    "    tfs = pd.Index(motif_bed[\"name\"]).intersection(rna.var_names)\n",
    "    rna[:, np.union1d(genes, tfs)].write_loom(f\"{work_dir}/rna.loom\")\n",
    "    np.savetxt(f\"{work_dir}/tfs.txt\", tfs, fmt=\"%s\")\n",
    "\n",
    "    # Construct the command \n",
    "    command = (\n",
    "        f\"pyscenic grn {work_dir}/rna.loom {work_dir}/tfs.txt \"\n",
    "        f\"-o {work_dir}/draft_grn.csv --seed 0 --num_workers {num_workers} \"\n",
    "        \"--cell_id_attribute obs_id --gene_attribute name\"\n",
    "    )\n",
    "\n",
    "    result = subprocess.run(command, shell=True, capture_output=True, text=True)\n",
    "\n",
    "    print(\"Output:\")\n",
    "    print(result.stdout)\n",
    "    print(\"Error:\")\n",
    "    print(result.stderr)\n",
    "\n",
    "    if result.returncode == 0:\n",
    "        print(\"Command executed successfully\")\n",
    "    else:\n",
    "        print(\"Command failed with return code\", result.returncode)\n",
    "\n",
    "\n",
    "    ## Generate TF cis-regulatory ranking bridged by ATAC peaks\n",
    "    peak_bed = scglue.genomics.Bed(atac.var.loc[peaks])\n",
    "    peak2tf = scglue.genomics.window_graph(peak_bed, motif_bed, 0, right_sorted=True)\n",
    "    peak2tf = peak2tf.edge_subgraph(e for e in peak2tf.edges if e[1] in tfs)\n",
    "\n",
    "    gene2tf_rank_glue = scglue.genomics.cis_regulatory_ranking(\n",
    "        gene2peak, peak2tf, genes, peaks, tfs,\n",
    "        region_lens=atac.var.loc[peaks, \"chromEnd\"] - atac.var.loc[peaks, \"chromStart\"],\n",
    "        random_state=0)\n",
    "\n",
    "    flank_bed = scglue.genomics.Bed(rna.var.loc[genes]).strand_specific_start_site().expand(500, 500)\n",
    "    flank2tf = scglue.genomics.window_graph(flank_bed, motif_bed, 0, right_sorted=True)\n",
    "\n",
    "    gene2flank = nx.Graph([(g, g) for g in genes])\n",
    "    gene2tf_rank_supp = scglue.genomics.cis_regulatory_ranking(\n",
    "        gene2flank, flank2tf, genes, genes, tfs,\n",
    "        n_samples=0\n",
    "    )\n",
    "\n",
    "    ### Prune coexpression network using cis-regulatory ranking\n",
    "\n",
    "    gene2tf_rank_glue.columns = gene2tf_rank_glue.columns + \"_glue\"\n",
    "    gene2tf_rank_supp.columns = gene2tf_rank_supp.columns + \"_supp\"\n",
    "\n",
    "    scglue.genomics.write_scenic_feather(gene2tf_rank_glue, f\"{work_dir}/glue.genes_vs_tracks.rankings.feather\")\n",
    "    scglue.genomics.write_scenic_feather(gene2tf_rank_supp, f\"{work_dir}/supp.genes_vs_tracks.rankings.feather\")\n",
    "\n",
    "    pd.concat([\n",
    "        pd.DataFrame({\n",
    "                \"#motif_id\": tfs + \"_glue\",\n",
    "                \"gene_name\": tfs\n",
    "            }),\n",
    "        pd.DataFrame({\n",
    "            \"#motif_id\": tfs + \"_supp\",\n",
    "            \"gene_name\": tfs\n",
    "        })\n",
    "    ]).assign(\n",
    "        motif_similarity_qvalue=0.0,\n",
    "        orthologous_identity=1.0,\n",
    "        description=\"placeholder\"\n",
    "    ).to_csv(f\"{work_dir}/ctx_annotation.tsv\", sep=\"\\t\", index=False)\n",
    "\n",
    "    # Construct the command \n",
    "    #TODO: be sure that obs_id is in obs and name is in var\n",
    "    command = (\n",
    "        f\" pyscenic ctx {work_dir}/draft_grn.csv {work_dir}/glue.genes_vs_tracks.rankings.feather \"\n",
    "        f\" {work_dir}/supp.genes_vs_tracks.rankings.feather  --annotations_fname {work_dir}/ctx_annotation.tsv \"\n",
    "        f\" --expression_mtx_fname {work_dir}/rna.loom --output {work_dir}/pruned_grn.csv \"\n",
    "        f\" --rank_threshold 500 --min_genes 1  --num_workers {num_workers} \"\n",
    "        \" --cell_id_attribute obs_id --gene_attribute name\"\n",
    "    )\n",
    "\n",
    "    result = subprocess.run(command, shell=True, capture_output=True, text=True)\n",
    "\n",
    "    print(\"Output:\")\n",
    "    print(result.stdout)\n",
    "    print(\"Error:\")\n",
    "    print(result.stderr)\n",
    "\n",
    "    if result.returncode == 0:\n",
    "        print(\"pyscenic ctx executed successfully\")\n",
    "    else:\n",
    "        print(\"pyscenic ctx failed with return code\", result.returncode)\n",
    "cis_inference_(num_workers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5898bb8f-391f-49c9-924b-d9aead6e1332",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-07-03T12:38:53.115924Z",
     "iopub.status.busy": "2024-07-03T12:38:53.115535Z",
     "iopub.status.idle": "2024-07-03T12:38:53.118823Z",
     "shell.execute_reply": "2024-07-03T12:38:53.118267Z",
     "shell.execute_reply.started": "2024-07-03T12:38:53.115898Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# pd.read_csv(f\"{work_dir}/pruned_grn.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "40e3242f-f522-4053-b484-a9878be2f5cf",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-07-02T19:22:07.542864Z",
     "iopub.status.busy": "2024-07-02T19:22:07.542461Z",
     "iopub.status.idle": "2024-07-02T19:22:07.564885Z",
     "shell.execute_reply": "2024-07-02T19:22:07.564066Z",
     "shell.execute_reply.started": "2024-07-02T19:22:07.542836Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>weight</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [source, target, weight]\n",
       "Index: []"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# grn = scglue.genomics.read_ctx_grn(\"pruned_grn.csv\")\n",
    "from ast import literal_eval\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\n",
    "    f\"{work_dir}/pruned_grn.csv\", header=None, skiprows=3,\n",
    "    usecols=[0, 8], names=[\"TF\", \"targets\"]\n",
    ")\n",
    "\n",
    "tfs_list = []\n",
    "target_list = []\n",
    "weight_list = []\n",
    "for i, (tf, targets) in df.iterrows():\n",
    "    for target, weight in literal_eval(targets):\n",
    "        tfs_list.append(tf)\n",
    "        target_list.append(target)\n",
    "        weight_list.append(weight)\n",
    "scglue_grn = pd.DataFrame(np.stack([tfs_list, target_list, weight_list], axis=1), columns=['source','target','weight'])\n",
    "scglue_grn.weight = scglue_grn.weight.astype(float)\n",
    "scglue_grn = scglue_grn.drop_duplicates().reset_index(drop=True)\n",
    "scglue_grn"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "saturn (Python 3)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
