{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/root/.local/lib/python3.10/site-packages/numba/np/ufunc/parallel.py:371: NumbaWarning: \u001b[1mThe TBB threading layer requires TBB version 2021 update 6 or later i.e., TBB_INTERFACE_VERSION >= 12060. Found TBB_INTERFACE_VERSION = 12050. The TBB threading layer is disabled.\u001b[0m\n",
      "  warnings.warn(problem)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os \n",
    "from celloracle import motif_analysis as ma\n",
    "import pandas as pd\n",
    "import celloracle as co\n",
    "import anndata\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "%matplotlib inline\n",
    "\n",
    "\n",
    "work_dir = '../../output'\n",
    "celloracle_dir = f'{work_dir}/celloracle'\n",
    "cicero_dir = f'{work_dir}/cicero'\n",
    "kaggle_data_dir = '../../input/kaggle/input'\n",
    "os.makedirs(f'{celloracle_dir}', exist_ok=True)\n",
    "os.makedirs(f'{cicero_dir}/scATAC', exist_ok=True)\n",
    "os.makedirs(f'{celloracle_dir}/baseGRN', exist_ok=True)\n",
    "os.makedirs(f'{celloracle_dir}/GRN', exist_ok=True)\n",
    "\n",
    "def format_peakname(df):\n",
    "    df.loc[:, 'location'] = df.location.str.replace(':','_')\n",
    "    df.loc[:, 'location'] = df.location.str.replace('-','_')\n",
    "    return  df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cicero"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Format the data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "multiome_var_meta = pd.read_csv(f'{kaggle_data_dir}/open-problems-single-cell-perturbations/multiome_var_meta.csv')\n",
    "multiome_train = pd.read_parquet(f'{kaggle_data_dir}/open-problems-single-cell-perturbations/multiome_train.parquet')\n",
    "multiome_obs_meta = pd.read_csv(f'{kaggle_data_dir}/open-problems-single-cell-perturbations/multiome_obs_meta.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract only peaks\n",
    "peaks_locations = multiome_var_meta[multiome_var_meta['feature_type']=='Peaks']['location'].values\n",
    "df_peaks = multiome_train[multiome_train['location'].isin(peaks_locations)].reset_index(drop=True)\n",
    "del multiome_train\n",
    "# extract peaks for donor and cell type\n",
    "# format and save\n",
    "df_peaks = format_peakname(df_peaks)\n",
    "df_peaks[['location','obs_id','count']].to_csv(ff'{work_dir}/cicero/scATAC/peaks.txt', sep='\\t', header=None, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get chromosome sizes (for hg38 here). \n",
    "target_url='http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.chrom.sizes'\n",
    "chromsizes=pd.read_csv(target_url, sep='\\t', header=None)\n",
    "chromsizes.columns=['Chromosome', 'End']\n",
    "chromsizes['Start']=[0]*chromsizes.shape[0]\n",
    "chromsizes=chromsizes.loc[:,['Chromosome', 'Start', 'End']]\n",
    "# Exceptionally in this case, to agree with CellRangerARC annotations\n",
    "chromsizes['Chromosome'] = [chromsizes['Chromosome'][x].replace('v', '.') for x in range(len(chromsizes['Chromosome']))]\n",
    "chromsizes['Chromosome'] = [chromsizes['Chromosome'][x].split('_')[1] if len(chromsizes['Chromosome'][x].split('_')) > 1 else chromsizes['Chromosome'][x] for x in range(len(chromsizes['Chromosome']))]\n",
    "\n",
    "chromsizes = chromsizes[['Chromosome', 'End']]\n",
    "# save\n",
    "chromsizes.to_csv(f'{work_dir}/cicero/chromsizes.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check if all chr are found\n",
    "chrs = df_peaks.locations.apply(lambda x:x.split('_')[0])\n",
    "chrs.isin(chromsizes.Chromosome).sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Actual run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "library(Gviz)\n",
    "library(GenomicRanges)\n",
    "library(rtracklayer)\n",
    "library(cicero)\n",
    "set.seed(2017)\n",
    "packageVersion(\"cicero\")\n",
    "\n",
    "# get the peaks\n",
    "cicero_data <- read.table(paste0(\"../output/cicero/scATAC/peaks.txt\"), sep = \"\\t\", header = FALSE)\n",
    "input_cds <- make_atac_cds(cicero_data, binarize = TRUE)\n",
    "input_cds <- monocle3::detect_genes(input_cds)\n",
    "input_cds <- estimate_size_factors(input_cds)\n",
    "input_cds <- preprocess_cds(input_cds, method = \"LSI\")\n",
    "input_cds <- reduce_dimension(input_cds, reduction_method = 'UMAP', preprocess_method = \"LSI\")\n",
    "# reduced dimension and cicero object\n",
    "umap_coords <- reducedDims(input_cds)$UMAP\n",
    "cicero_cds <- make_cicero_cds(input_cds, reduced_coordinates = umap_coords)\n",
    "# read chromsize\n",
    "chromsizes <- read.csv(f'{work_dir}/cicero/chromsizes.csv')\n",
    "# actual run\n",
    "conns <- run_cicero(cicero_cds, chromsizes)\n",
    "# save all peaks and connections\n",
    "all_peaks <- row.names(exprs(input_cds))\n",
    "write.csv(x = all_peaks, file = \"../output/cicero/all_peaks.csv\")\n",
    "write.csv(x = conns, file = \"../output/cicero/connections.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Celloracle"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Base GRN\n",
    "This section create base GRN given cicero results.\n",
    "###  Associate peaks with TSS\n",
    "Each peak is associated with promotors of target genes (+- 1kbp). Celloracle should be installed for this step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "que bed peaks: 131047\n",
      "tss peaks in que: 20898\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "***** WARNING: File /tmp/pybedtools.f78i1zv8.tmp has inconsistent naming convention for record:\n",
      "GL000194.1\t55749\t56580\n",
      "\n",
      "***** WARNING: File /tmp/pybedtools.f78i1zv8.tmp has inconsistent naming convention for record:\n",
      "GL000194.1\t55749\t56580\n",
      "\n"
     ]
    }
   ],
   "source": [
    "peaks = pd.read_csv(f\"{cicero_dir}/all_peaks.csv\", index_col=0)\n",
    "cicero_connections =  pd.read_csv(f\"{cicero_dir}/connections.csv\", index_col=0)\n",
    "tss_annotated = ma.get_tss_info(peak_str_list=peaks['x'].values, ref_genome=\"hg38\")\n",
    "tss_annotated.to_csv(f'{work_dir}/celloracle/baseGRN/tss_annotated.csv')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  Merge TSS peaks with cicero connections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "integrated = ma.integrate_tss_peak_with_cicero(tss_peak=tss_annotated, \n",
    "                                               cicero_connections=cicero_connections)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "integrated.to_csv(f'{work_dir}/infer/celloracle/baseGRN/integrated.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  Shortlisten the peak-gene connections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>peak_id</th>\n",
       "      <th>gene_short_name</th>\n",
       "      <th>coaccess</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10_100009475_100010367</td>\n",
       "      <td>DNMBP</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10_100185622_100186542</td>\n",
       "      <td>ERLIN1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10_100229082_100229931</td>\n",
       "      <td>CHUK</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10_100267238_100268104</td>\n",
       "      <td>CWF19L1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10_100286187_100287039</td>\n",
       "      <td>BLOC1S2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19109</th>\n",
       "      <td>chrX_9464670_9465539</td>\n",
       "      <td>TBL1X</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19110</th>\n",
       "      <td>chrX_96684366_96685202</td>\n",
       "      <td>DIAPH2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19111</th>\n",
       "      <td>chrX_97564682_97565222</td>\n",
       "      <td>DIAPH2-AS1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19112</th>\n",
       "      <td>chrY_13703349_13704256</td>\n",
       "      <td>TMSB4Y</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19113</th>\n",
       "      <td>chrY_20575222_20576136</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19114 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         peak_id gene_short_name  coaccess\n",
       "0      chr10_100009475_100010367           DNMBP       1.0\n",
       "1      chr10_100185622_100186542          ERLIN1       1.0\n",
       "2      chr10_100229082_100229931            CHUK       1.0\n",
       "3      chr10_100267238_100268104         CWF19L1       1.0\n",
       "4      chr10_100286187_100287039         BLOC1S2       1.0\n",
       "...                          ...             ...       ...\n",
       "19109       chrX_9464670_9465539           TBL1X       1.0\n",
       "19110     chrX_96684366_96685202          DIAPH2       1.0\n",
       "19111     chrX_97564682_97565222      DIAPH2-AS1       1.0\n",
       "19112     chrY_13703349_13704256          TMSB4Y       1.0\n",
       "19113     chrY_20575222_20576136          EIF1AY       1.0\n",
       "\n",
       "[19114 rows x 3 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peak_gene = integrated[integrated.coaccess>0.8].reset_index(drop=True)\n",
    "peak_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "peak_gene.to_csv(f'{work_dir}/infer/celloracle/baseGRN/peak_gene.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create base GRN\n",
    "Running interactively might take a long time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PLEASE make sure reference genome is correct.\n",
    "ref_genome = \"hg38\"\n",
    "\n",
    "genome_installation = ma.is_genome_installed(ref_genome=ref_genome,\n",
    "                                             genomes_dir=None)\n",
    "print(ref_genome, \"installation: \", genome_installation)\n",
    "\n",
    "# Instantiate TFinfo object\n",
    "tfi = ma.TFinfo(peak_data_frame=peak_gene, \n",
    "                ref_genome=\"hg38\",\n",
    "                genomes_dir=None) \n",
    "\n",
    "tfi.scan(fpr=0.05, \n",
    "         motifs=None,  # If you enter None, default motifs will be loaded.\n",
    "         verbose=True)\n",
    "# Check motif scan results\n",
    "tfi.scanned_df.head()\n",
    "# Reset filtering \n",
    "tfi.reset_filtering()\n",
    "\n",
    "# Do filtering\n",
    "tfi.filter_motifs_by_score(threshold=10)\n",
    "\n",
    "# Format post-filtering results.\n",
    "tfi.make_TFinfo_dataframe_and_dictionary(verbose=True)\n",
    "\n",
    "# Format and save \n",
    "df = tfi.to_dataframe()\n",
    "df.head()\n",
    "df.to_csv(f'{work_dir}/infer/celloracle/baseGRN/grn_celloracle_base.csv')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Base GRN: only proximal \n",
    "This is only based on proximal cis elements and doesnt require cicero"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata as ad \n",
    "adata_atac = ad.read_h5ad(f'{work_dir}/scATAC/adata_atac.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "peaks = format_peakname(adata_atac.var.reset_index()).location.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "que bed peaks: 135418\n",
      "tss peaks in que: 21028\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "***** WARNING: File /tmp/pybedtools.dxm99xoa.tmp has inconsistent naming convention for record:\n",
      "chr10\t100001032\t100001800\n",
      "\n",
      "***** WARNING: File /tmp/pybedtools.dxm99xoa.tmp has inconsistent naming convention for record:\n",
      "chr10\t100001032\t100001800\n",
      "\n"
     ]
    }
   ],
   "source": [
    "tss_annotated = ma.get_tss_info(peak_str_list=peaks, ref_genome=\"hg38\")\n",
    "tss_annotated.to_csv(f'{work_dir}/infer/celloracle/baseGRN/tss_annotated.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "##----- integrate_tss_peak_with_cicero\n",
    "import numpy as np\n",
    "from celloracle.motif_analysis.process_bed_file import df_to_list_peakstr\n",
    "# 1. check tss data format and convert if needed\n",
    "tss_peak=tss_annotated\n",
    "tss = tss_peak.copy()\n",
    "if np.all([i in tss.columns for i in [\"chr\", \"start\", \"end\"]]):\n",
    "    tss = pd.DataFrame({\"peak_id\": df_to_list_peakstr(tss),\n",
    "                        \"gene_short_name\": tss.gene_short_name.values})\n",
    "else:\n",
    "    raise ValueError(\"tss_peak format error\")\n",
    "\n",
    "peak_gene = tss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "peak_gene.to_csv(f'{work_dir}/infer/celloracle/baseGRN/peak_gene.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TF motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "from celloracle import motif_analysis as ma\n",
    "import genomepy\n",
    "genomes_dir='/beegfs/desy/user/nourisaj/op_multiomics_grn/output/celloracle'\n",
    "peak_gene = pd.read_csv(f'{work_dir}/infer/celloracle/baseGRN/peak_gene.csv', index_col=0)\n",
    "genomepy.install_genome(name=\"hg38\", provider=\"UCSC\", genomes_dir=genomes_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PLEASE make sure reference genome is correct.\n",
    "ref_genome = \"hg38\"\n",
    "\n",
    "genome_installation = ma.is_genome_installed(ref_genome=ref_genome,\n",
    "                                             genomes_dir=genomes_dir)\n",
    "print(ref_genome, \"installation: \", genome_installation)\n",
    "\n",
    "# Instantiate TFinfo object\n",
    "tfi = ma.TFinfo(peak_data_frame=peak_gene, \n",
    "                ref_genome=\"hg38\",\n",
    "                genomes_dir=genomes_dir) \n",
    "\n",
    "tfi.scan(fpr=0.05, \n",
    "         motifs=None,  # If you enter None, default motifs will be loaded.\n",
    "         verbose=True)\n",
    "# Check motif scan results\n",
    "tfi.scanned_df.head()\n",
    "# Reset filtering \n",
    "tfi.reset_filtering()\n",
    "\n",
    "# Do filtering\n",
    "tfi.filter_motifs_by_score(threshold=10)\n",
    "\n",
    "# Format post-filtering results.\n",
    "tfi.make_TFinfo_dataframe_and_dictionary(verbose=True)\n",
    "\n",
    "# Format and save \n",
    "df = tfi.to_dataframe()\n",
    "df.head()\n",
    "df.to_csv(f'{work_dir}/infer/celloracle/baseGRN/grn_celloracle_base.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GRN construction\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preprocessing scRNA-seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import celloracle as co\n",
    "import anndata\n",
    "import scanpy as sc\n",
    "adata = anndata.read_h5ad(f'{work_dir}/scRNA/adata_rna.h5ad')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 25551 × 22787\n",
       "    obs: 'cell_type', 'donor_id', 'n_genes'\n",
       "    var: 'n_cells'\n",
       "    layers: 'counts', 'x_norm'"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "del adata.varm \n",
    "del adata.uns \n",
    "del adata.obsp \n",
    "del adata.obsm \n",
    "del adata.obs['louvain']\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X = adata.layers['counts'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1., 1., 1., ..., 2., 1., 1.])"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_per_cell(adata, key_n_counts='n_counts_all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "filter_result = sc.pp.filter_genes_dispersion(adata.X,\n",
    "                                              flavor='cell_ranger',\n",
    "                                              n_top_genes=3000,\n",
    "                                              log=False)\n",
    "\n",
    "# Subset the genes\n",
    "adata = adata[:, filter_result.gene_subset]\n",
    "\n",
    "# Renormalize after filtering\n",
    "sc.pp.normalize_per_cell(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Log transformation and scaling\n",
    "sc.pp.log1p(adata)\n",
    "sc.pp.scale(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PCA\n",
    "sc.tl.pca(adata, svd_solver='arpack')\n",
    "\n",
    "# Diffusion map\n",
    "sc.pp.neighbors(adata, n_neighbors=4, n_pcs=20)\n",
    "\n",
    "sc.tl.diffmap(adata)\n",
    "# Calculate neihbors again based on diffusionmap \n",
    "sc.pp.neighbors(adata, n_neighbors=10, use_rep='X_diffmap')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.louvain(adata, resolution=0.8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 25551 × 3000\n",
       "    obs: 'cell_type', 'donor_id', 'n_genes', 'n_counts_all', 'n_counts', 'louvain'\n",
       "    var: 'n_cells', 'mean', 'std'\n",
       "    uns: 'log1p', 'pca', 'neighbors', 'diffmap_evals', 'louvain'\n",
       "    obsm: 'X_pca', 'X_diffmap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'counts', 'x_norm'\n",
       "    obsp: 'distances', 'connectivities'"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.paga(adata, groups='louvain')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.paga(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Package 'fa2' is not installed, falling back to layout 'fr'.To use the faster and better ForceAtlas2 layout, install package 'fa2' (`pip install fa2`).\n"
     ]
    }
   ],
   "source": [
    "sc.tl.draw_graph(adata, init_pos='paga', random_state=123)\n",
    "sc.pl.draw_graph(adata, color='louvain', legend_loc='on data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Metadata columns : ['cell_type', 'donor_id', 'n_genes', 'louvain', 'n_counts_all', 'n_counts']\n",
      "Dimensional reduction:  ['X_diffmap', 'X_draw_graph_fr', 'X_pca', 'X_umap']\n"
     ]
    }
   ],
   "source": [
    "# Check data in anndata\n",
    "print(\"Metadata columns :\", list(adata.obs.columns))\n",
    "print(\"Dimensional reduction: \", list(adata.obsm.keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1., 1., 1., ..., 1., 1., 2.])"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## run based on counts as suggesyed by co pipeline\n",
    "adata.X = adata.layers[\"counts\"]\n",
    "adata.X.data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load base GRN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_GRN = pd.read_csv(f'{work_dir}/infer/celloracle/baseGRN/grn_celloracle_base.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Initialize "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: adata.X seems to be already log-transformed.\n"
     ]
    }
   ],
   "source": [
    "# Instantiate Oracle object\n",
    "oracle = co.Oracle()\n",
    "# Instantiate Oracle object.\n",
    "oracle.import_anndata_as_raw_count(adata=adata,\n",
    "                                   cluster_column_name=\"cell_type\",\n",
    "                                   embedding_name=\"X_draw_graph_fr\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "# You can load TF info dataframe with the following code.\n",
    "oracle.import_TF_data(TF_info_matrix=base_GRN)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "oracle.perform_PCA()\n",
    "plt.plot(np.cumsum(oracle.pca.explained_variance_ratio_)[:100])\n",
    "n_comps = np.where(np.diff(np.diff(np.cumsum(oracle.pca.explained_variance_ratio_))>0.002))[0][0]\n",
    "plt.axvline(n_comps, c=\"k\")\n",
    "plt.show()\n",
    "print(n_comps)\n",
    "n_comps = min(n_comps, 50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cell number is :25551\n",
      "Auto-selected k is :50\n"
     ]
    }
   ],
   "source": [
    "n_cell = oracle.adata.shape[0]\n",
    "print(f\"cell number is :{n_cell}\")\n",
    "k = min([int(0.025*n_cell), 50])\n",
    "print(f\"Auto-selected k is :{k}\")\n",
    "oracle.knn_imputation(n_pca_dims=n_comps, k=k, balanced=True, b_sight=k*8,\n",
    "                      b_maxl=k*4, n_jobs=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "oracle.to_hdf5(f'{work_dir}/infer/celloracle/grn/presaved_3000.celloracle.oracle')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GRN calculation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if True: # run locally\n",
    "    # Load file.\n",
    "    oracle = co.load_hdf5(f'{work_dir}/infer/celloracle/grn/presaved_3000.celloracle.oracle')\n",
    "    # This step may take some time.\n",
    "    links = oracle.get_links(cluster_name_for_GRN_unit=\"cell_type\", alpha=10,\n",
    "                            verbose_level=10)\n",
    "    links.to_hdf5(file_path=f\"{work_dir}/infer/celloracle/grn/links_3000.celloracle.links\")\n",
    "else:\n",
    "    !python celloracle/run_grn.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Post evaluation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Peak gene connections\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21028\n",
      "368203\n",
      "21028\n"
     ]
    }
   ],
   "source": [
    "tss_annotated = pd.read_csv(f'{work_dir}/infer/celloracle/baseGRN/tss_annotated.csv')\n",
    "print(len(tss_annotated))\n",
    "integrated = pd.read_csv(f'{work_dir}/infer/celloracle/baseGRN/integrated.csv')\n",
    "print(len(integrated))\n",
    "peak_gene_shortlist = pd.read_csv(f'{work_dir}/infer/celloracle/baseGRN/peak_gene.csv')\n",
    "print(len(peak_gene_shortlist))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "unique peaks in peak-gene (17295,)\n",
      "unique genes in peak-gene (16691,)\n"
     ]
    }
   ],
   "source": [
    "print('unique peaks in peak-gene', peak_gene_shortlist.peak_id.unique().shape)\n",
    "print('unique genes in peak-gene', peak_gene_shortlist.gene_short_name.unique().shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of DORC genes with t of 10  0\n",
      "number of DORC genes with t of 5  17\n"
     ]
    }
   ],
   "source": [
    "peak_gene_shortlist = pd.read_csv(f'{work_dir}/infer/celloracle/baseGRN/peak_gene.csv', index_col=0)\n",
    "peak_gene_co_n = peak_gene_shortlist.groupby('gene_short_name').apply(lambda df:df['peak_id'].shape[0])\n",
    "np.max(peak_gene_co_n.values), np.median(peak_gene_co_n.values)\n",
    "\n",
    "# print('number of TFs ', scenicplus.TF.unique().shape[0], ' CIS ', scenicplus.Region.unique().shape[0], ' gene ', scenicplus.Gene.unique().shape[0])\n",
    "print('number of DORC genes with t of 10 ', (peak_gene_co_n.values > 10).sum())\n",
    "print('number of DORC genes with t of 5 ', (peak_gene_co_n.values > 5).sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert peak to peak_id using celloracle function\n",
    "tss_annotated_df = pd.DataFrame({\"peak_id\": ma.process_bed_file.df_to_list_peakstr(tss_annotated),\n",
    "                            \"gene_short_name\": tss_annotated.gene_short_name.values})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# number of tss_annotated pairs in shortlisted peak\n",
    "print(f'Percentage of proximal elements in the final peak gene pairs: {100*peak_gene_shortlist.peak_id.isin(tss_annotated_df.peak_id).sum()/len(peak_gene_shortlist)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dorc_shortlisted = peak_gene_shortlist.groupby('gene_short_name').size()\n",
    "print(f\"In the short list: max peaks per gene: {dorc_shortlisted.max()}, median: {dorc_shortlisted.median()}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "threshold = 10\n",
    "peak_new = integrated[integrated.coaccess >= threshold]\n",
    "print('number of DORC: ', (peak_new.groupby('gene_short_name').size()>10).sum())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Base GRN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_GRN = pd.read_csv(f'{work_dir}/infer/celloracle/baseGRN/grn_celloracle_base.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tfs_co = base_GRN.columns[3:]\n",
    "keeo_cols = tfs_co.insert(0, ['gene_short_name', 'peak_id'])\n",
    "df = base_GRN[keeo_cols]\n",
    "# Melting the DataFrame\n",
    "melted_df = pd.melt(df, id_vars=['gene_short_name', 'peak_id'], var_name='TF', value_name='Link')\n",
    "\n",
    "# Filtering out rows where there is no link (optional, if you only want interactions)\n",
    "melted_df = melted_df[melted_df['Link'] == 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f'TFs : {melted_df.TF.unique().shape} , regions : {melted_df.peak_id.unique().shape},  genes : {melted_df.gene_short_name.unique().shape}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Refined GRN "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "tag = '' #'_hvg'\n",
    "\n",
    "links = co.load_hdf5(f\"{work_dir}/infer/celloracle/grn/links{tag}.celloracle.links\") \n",
    "links.filter_links(p=0.05, weight=\"coef_abs\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TFs : (596,) , genes : (13845,)\n"
     ]
    }
   ],
   "source": [
    "links_stack = []\n",
    "cell_types = list(links.links_dict.keys())\n",
    "links_celltype = links.links_dict[cell_types[2]]\n",
    "print(f'TFs : {links_celltype.source.unique().shape} , genes : {links_celltype.target.unique().shape}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of links before top quantile selection:  2396360\n",
      "connections:  (50000, 4)\n",
      "TFs:  (477,)\n",
      "target gene:  (5885,)\n",
      "number of links before top quantile selection:  2392742\n",
      "connections:  (50000, 4)\n",
      "TFs:  (467,)\n",
      "target gene:  (6341,)\n",
      "number of links before top quantile selection:  2388828\n",
      "connections:  (50000, 4)\n",
      "TFs:  (460,)\n",
      "target gene:  (5636,)\n",
      "number of links before top quantile selection:  2399961\n",
      "connections:  (50000, 4)\n",
      "TFs:  (419,)\n",
      "target gene:  (6973,)\n"
     ]
    }
   ],
   "source": [
    "top_links = 50_000\n",
    "grn_stack = []\n",
    "for i, cell_type in enumerate(cell_types):\n",
    "    grn_celltype = links.links_dict[cell_type]\n",
    "    grn_celltype = grn_celltype[~(grn_celltype.coef_mean==0)][['source', 'target', 'coef_mean']] # remove those with 0 coeff\n",
    "    print('number of links before top quantile selection: ', grn_celltype.shape[0])\n",
    "    grn_celltype.columns = ['source', 'target', f'weight']\n",
    "    grn_celltype['cell_type'] = cell_type\n",
    "    grn_celltype = grn_celltype.sort_values(by='weight', key=abs, ascending=False).head(top_links)\n",
    "    print('connections: ', grn_celltype.shape)\n",
    "    print('TFs: ', grn_celltype.source.unique().shape)\n",
    "    print('target gene: ', grn_celltype.target.unique().shape)\n",
    "    grn_stack.append(grn_celltype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>weight</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STAT1</td>\n",
       "      <td>RPS26</td>\n",
       "      <td>-0.308208</td>\n",
       "      <td>B cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SPI1</td>\n",
       "      <td>RPS18</td>\n",
       "      <td>0.289873</td>\n",
       "      <td>B cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BCL11B</td>\n",
       "      <td>LEF1</td>\n",
       "      <td>0.289265</td>\n",
       "      <td>B cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPI1</td>\n",
       "      <td>RPL32</td>\n",
       "      <td>0.284154</td>\n",
       "      <td>B cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STAT1</td>\n",
       "      <td>XIST</td>\n",
       "      <td>-0.278137</td>\n",
       "      <td>B cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199995</th>\n",
       "      <td>ETS1</td>\n",
       "      <td>LINC02273</td>\n",
       "      <td>-0.018031</td>\n",
       "      <td>agg_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199996</th>\n",
       "      <td>EGR1</td>\n",
       "      <td>SRSF3</td>\n",
       "      <td>-0.018030</td>\n",
       "      <td>agg_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199997</th>\n",
       "      <td>YY1</td>\n",
       "      <td>CYBA</td>\n",
       "      <td>0.018029</td>\n",
       "      <td>agg_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199998</th>\n",
       "      <td>FOS</td>\n",
       "      <td>PTBP1</td>\n",
       "      <td>0.018029</td>\n",
       "      <td>agg_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199999</th>\n",
       "      <td>PAX5</td>\n",
       "      <td>ELOC</td>\n",
       "      <td>0.018029</td>\n",
       "      <td>agg_type</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>200000 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        source     target    weight cell_type\n",
       "0        STAT1      RPS26 -0.308208   B cells\n",
       "1         SPI1      RPS18  0.289873   B cells\n",
       "2       BCL11B       LEF1  0.289265   B cells\n",
       "3         SPI1      RPL32  0.284154   B cells\n",
       "4        STAT1       XIST -0.278137   B cells\n",
       "...        ...        ...       ...       ...\n",
       "199995    ETS1  LINC02273 -0.018031  agg_type\n",
       "199996    EGR1      SRSF3 -0.018030  agg_type\n",
       "199997     YY1       CYBA  0.018029  agg_type\n",
       "199998     FOS      PTBP1  0.018029  agg_type\n",
       "199999    PAX5       ELOC  0.018029  agg_type\n",
       "\n",
       "[200000 rows x 4 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "co_grn = pd.concat(grn_stack).reset_index(drop=True)\n",
    "co_grn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjkAAAGdCAYAAADwjmIIAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/P9b71AAAACXBIWXMAAA9hAAAPYQGoP6dpAAA0o0lEQVR4nO3deXRV5b3/8Q8JiaABCkkopeCVm5hBMh2gNyY9NC0KWJBeQ2SoRIGmhVogWAfgR1GI0gQQvRi1hQJGg7Ep1xDUAr2rWuSGlcSBJmAoU0AQm2oGhHCYMu3fHy725RARTgiGPHm/1spanGd/997P87jN+WQP53SyLMsSAACAYbzaugMAAADXAiEHAAAYiZADAACMRMgBAABGIuQAAAAjEXIAAICRCDkAAMBIhBwAAGCkzm3dgbbU1NSkyspK3XTTTerUqVNbdwcAAFwBy7J06tQp9e7dW15elz5f06FDTmVlpRISEtq6GwAAoAW2bdumPn36XHJ5hw45N910k6QvJ8nPz6+NewMAAK6Ey+VSQkKC/T5+KR065Jy/ROXn50fIAQCgnbncrSbceAwAAIxEyAEAAEYi5AAAACMRcgAAgJEIOQAAwEiEHAAAYCRCDgAAMBIhBwAAGImQAwAAjETIAQAARiLkAAAAIxFyAACAkQg5AADASIQcAABgJEIOgOtOY5P1ta8B4Ep0busOAMDFvL06aXZuicorXQru7afnJjrauksA2iFCDoDrUnmlS7sratu6GwDaMS5XAQAAIxFyAACAkQg5AADASIQcAABgJEIOAAAwEiEHAAAYiZADAACMRMgBAABGIuQAAAAjEXIAAICRCDkAAMBIhBwAAGAkj0LOsGHDFBoa2uwnLS1NknTu3DmlpaUpNjZWDodDs2bNUnV1tds2KioqNG3aNEVHRysuLk5Lly5VQ0ODW817772nxMRERUREaPjw4dqwYUOzvuTk5GjYsGGKjIzUuHHjtGvXLk/HDgAADOZRyHn99de1fft2+ycrK0uSdNddd0mS0tPTtXXrVq1YsULr1q1TZWWlZs6caa/f2Nio6dOnq76+Xrm5uVqyZIny8/OVmZlp1xw9elTTp09XbGys3njjDU2ePFkLFixQQUGBXbN582ZlZGRoxowZys/PV1hYmFJSUlRTU3NVkwEAAMzhUcjp1auXAgMD7Z+tW7fq5ptv1n/8x3/o5MmTysvL07x58xQXF6eIiAilp6erpKREpaWlkqTt27ervLxcTz/9tMLDw5WQkKDZs2crJydHdXV1kqTc3Fz169dP8+bNU1BQkJKTkzVy5Ei9/PLLdj+ysrI0fvx4JSUlKTg4WGlpaerSpYvy8vJabWIAAED71uJ7curq6vTmm28qKSlJnTp1UllZmerr6xUfH2/XBAUFqW/fvnbIKS0tVUhIiAICAuwap9Mpl8ul8vJyuyYuLs5tX06n095GXV2ddu/e7bYfLy8vxcfHq6SkpKXDAQAAhmlxyHn77bd18uRJJSYmSpKqq6vl4+Oj7t27u9X5+/urqqrKrrkw4EiyX1+uxuVy6ezZs/riiy/U2Ngof3//Zvu5+P4fAADQcbU45OTl5ekHP/iBvv3tb7dmfwAAAFpFi0LOP//5TxUWFuree++12wICAlRfX6/a2lq32pqaGgUGBto1F59tOf/6cjV+fn7q0qWLevbsKW9v72Y3GdfU1DQ7AwQAADquFoWcDRs2yN/fXz/84Q/ttoiICPn4+KioqMhuO3TokCoqKhQTEyNJiomJ0f79+90CSmFhofz8/BQcHGzXFBcXu+2vsLDQ3oavr68GDhzotp+mpiYVFRXJ4XC0ZDgAAMBAHoecpqYmbdiwQffcc486d+5st3fr1k1JSUlasmSJiouLVVZWpvnz58vhcNgBxel0Kjg4WHPmzNHevXtVUFCgFStWaNKkSfL19ZUkTZw4UUePHtWyZct08OBB5eTkaMuWLZoyZYq9r6lTp2r9+vXKz8/XwYMHtWjRIp05c0Zjx469utkAAADG6Hz5EneFhYWqqKhQUlJSs2Xz58+Xl5eXUlNTVVdXJ6fTqYULF9rLvb29tXLlSi1atEgTJkxQ165dlZiYqNTUVLumf//+WrVqlTIyMpSdna0+ffpo8eLFGjp0qF0zatQoHTt2TJmZmaqqqlJ4eLjWrFnD5SoAAGDrZFmW1dadaCsul0uDBw/Wjh075Ofn19bdAXCB0ZkF2l1Rq4F9u2tT6tDLrwCgw7jS92++uwoAABiJkAMAAIxEyAEAAEYi5AAAACMRcgAAgJEIOQAAwEiEHAAAYCRCDgAAMBIhBwAAGImQAwAAjETIAQAARiLkAAAAIxFyAACAkQg5AADASIQcAABgJEIOAAAwEiEHAAAYiZADAACMRMgBAABGIuQAAAAjEXIAAICRCDkAAMBIhBwAAGAkQg4AADASIQcAABiJkAMAAIxEyAEAAEYi5AAAACMRcgAAgJEIOQAAwEiEHAAAYCRCDgAAMBIhBwAAGImQAwAAjETIAQAARiLkAAAAIxFyAACAkTwOOZ9//rkeffRRxcbGKioqSmPGjNFHH31kL7csS88995ycTqeioqI0ZcoUHT582G0bx48f1yOPPKJBgwZpyJAhmj9/vk6dOuVWs3fvXt13332KjIxUQkKCVq9e3awvW7Zs0V133aXIyEiNGTNG27Zt83Q4AADAUB6FnBMnTuinP/2pfHx8tHr1am3atElz585Vjx497JrVq1dr3bp1WrRokdavX6+uXbsqJSVF586ds2seffRRlZeXKysrSytXrtSHH36oJ554wl7ucrmUkpKivn37asOGDZozZ45eeOEF/elPf7Jr/v73v+uRRx7Rvffeq40bN+qOO+7QjBkztH///quZDwAAYAiPQs7q1avVp08fZWRkKCoqSv3795fT6dTNN98s6cuzONnZ2XrwwQd15513KiwsTMuWLVNlZaXefvttSdLBgwdVUFCgxYsXKzo6WkOGDNGCBQu0adMmff7555KkN998U/X19UpPT9ett96q0aNH6/7771dWVpbdl+zsbA0dOlQ///nPFRQUpIceeki33XabXn311daaGwAA0I55FHL+9re/KSIiQqmpqYqLi9M999yj9evX28s//fRTVVVVKT4+3m7r1q2boqOjVVJSIkkqKSlR9+7dFRkZadfEx8fLy8tLu3btkiSVlpZqyJAh8vX1tWucTqc+/vhjnThxwq6Ji4tz65/T6VRpaaknQwIAAIbyKOQcPXpUf/zjH3XLLbdo7dq1+ulPf6rFixcrPz9fklRVVSVJ8vf3d1vP399f1dXVkqTq6mr16tXLbXnnzp3Vo0cPe/3q6moFBAS41Zx/feF2Lq65cD8AAKBj6+xJsWVZioiI0MMPPyxJuu2223TgwAHl5uYqMTHxmnQQAACgJTw6kxMYGKigoCC3tn//939XRUWFvVySampq3Gpqamrssy4BAQE6duyY2/KGhgadOHHCXj8gIKDZGZnzry/czsU1F+4HAAB0bB6FnEGDBunjjz92azt8+LC++93vSpL69eunwMBAFRUV2ctdLpd27twph8MhSXI4HKqtrVVZWZldU1xcrKamJkVFRUmSYmJi9OGHH6q+vt6uKSws1IABA+wnuWJiYlRcXOzWl8LCQsXExHgyJAAAYCiPQs7kyZO1c+dOrVy5UkeOHNFbb72l9evX67777pMkderUSQ888IB+//vf65133tG+ffs0Z84c9e7dW3feeackKSgoSEOHDtXjjz+uXbt2aceOHXrqqac0evRoffvb35YkjRkzRj4+PvrNb36jAwcOaPPmzcrOztbUqVPtvjzwwAMqKCjQSy+9pIMHD+r5559XWVmZkpOTW2tuAABAO9bJsizLkxW2bt2qZ599VocPH1a/fv00depUjR8/3l5uWZYyMzO1fv161dbWavDgwVq4cKEGDBhg1xw/flxPPfWU/va3v8nLy0sjRozQggULdNNNN9k1e/fu1ZNPPqmPPvpIPXv2VHJysqZNm+bWly1btmjFihX65z//qVtuuUWPPfaYEhISrngsLpdLgwcP1o4dO+Tn5+fJNAC4xkZnFmh3Ra0G9u2uTalD27o7AK4jV/r+7XHIMQkhB7h+EXIAXMqVvn/z3VUAAMBIhBwAAGAkQg4AADASIQcAABiJkAMAAIxEyAEAAEYi5AAAACMRcgAAgJEIOQAAwEiEHAAAYCRCDgAAMBIhBwAAGImQAwAAjETIAQAARiLkAAAAIxFyAACAkQg5AADASIQcAABgJEIOAAAwEiEHAAAYiZADAACMRMgBAABGIuQAAAAjEXIAAICRCDkAAMBIhBwAAGAkQg4AADASIQcAABiJkAMAAIxEyAEAAEYi5AAAACMRcgAAgJEIOQAAwEiEHAAAYCRCDgAAMBIhBwAAGMmjkPP8888rNDTU7eeuu+6yl587d05paWmKjY2Vw+HQrFmzVF1d7baNiooKTZs2TdHR0YqLi9PSpUvV0NDgVvPee+8pMTFRERERGj58uDZs2NCsLzk5ORo2bJgiIyM1btw47dq1y5OhAAAAw3l8JufWW2/V9u3b7Z/XXnvNXpaenq6tW7dqxYoVWrdunSorKzVz5kx7eWNjo6ZPn676+nrl5uZqyZIlys/PV2Zmpl1z9OhRTZ8+XbGxsXrjjTc0efJkLViwQAUFBXbN5s2blZGRoRkzZig/P19hYWFKSUlRTU1NS+cBAAAYxuOQ4+3trcDAQPunV69ekqSTJ08qLy9P8+bNU1xcnCIiIpSenq6SkhKVlpZKkrZv367y8nI9/fTTCg8PV0JCgmbPnq2cnBzV1dVJknJzc9WvXz/NmzdPQUFBSk5O1siRI/Xyyy/bfcjKytL48eOVlJSk4OBgpaWlqUuXLsrLy7v6GQEAAEbwOOQcOXJETqdTd9xxhx555BFVVFRIksrKylRfX6/4+Hi7NigoSH379rVDTmlpqUJCQhQQEGDXOJ1OuVwulZeX2zVxcXFu+3Q6nfY26urqtHv3brf9eHl5KT4+XiUlJZ4OBwAAGKqzJ8VRUVHKyMjQgAEDVFVVpRdffFGTJk3SW2+9perqavn4+Kh79+5u6/j7+6uqqkqSVF1d7RZwJNmvL1fjcrl09uxZnThxQo2NjfL392+2n0OHDnkyHAAAYDCPQk5CQoL977CwMEVHR+tHP/qRtmzZoi5durR65wAAAFrqqh4h7969u2655RZ98sknCggIUH19vWpra91qampqFBgYKOnLMzIXP211/vXlavz8/NSlSxf17NlT3t7ezW4yrqmpaXYGCAAAdFxXFXJOnTqlo0ePKjAwUBEREfLx8VFRUZG9/NChQ6qoqFBMTIwkKSYmRvv373cLKIWFhfLz81NwcLBdU1xc7LafwsJCexu+vr4aOHCg236amppUVFQkh8NxNcMBAAAG8SjkLF26VO+//74+/fRT/f3vf9fMmTPl5eWlu+++W926dVNSUpKWLFmi4uJilZWVaf78+XI4HHZAcTqdCg4O1pw5c7R3714VFBRoxYoVmjRpknx9fSVJEydO1NGjR7Vs2TIdPHhQOTk52rJli6ZMmWL3Y+rUqVq/fr3y8/N18OBBLVq0SGfOnNHYsWNbbWIAAED75tE9OZ999pkefvhhHT9+XL169dLgwYO1fv16+zHy+fPny8vLS6mpqaqrq5PT6dTChQvt9b29vbVy5UotWrRIEyZMUNeuXZWYmKjU1FS7pn///lq1apUyMjKUnZ2tPn36aPHixRo6dKhdM2rUKB07dkyZmZmqqqpSeHi41qxZw+UqAABg62RZltXWnWgrLpdLgwcP1o4dO+Tn59fW3QFwgdGZBdpdUauBfbtrU+rQy68AoMO40vdvvrsKAAAYiZADAACMRMgBAABGIuQAAAAjEXIAAICRCDkAAMBIhBwAAGAkQg4AADASIQcAABiJkAMAAIxEyAEAAEYi5AAAACMRcgAAgJEIOQAAwEiEHAAAYCRCDgAAMBIhBwAAGImQAwAAjETIAQAARiLkAAAAIxFyAACAkQg5AADASIQcAABgJEIOAAAwEiEHAAAYiZADAACMRMgBAABGIuQAAAAjEXIAAICRCDkAAMBIhBwAAGAkQg4AADASIQcAABiJkAMAAIxEyAEAAEYi5AAAACMRcgAAgJGuKuT84Q9/UGhoqH7729/abefOnVNaWppiY2PlcDg0a9YsVVdXu61XUVGhadOmKTo6WnFxcVq6dKkaGhrcat577z0lJiYqIiJCw4cP14YNG5rtPycnR8OGDVNkZKTGjRunXbt2Xc1wAACAQVoccnbt2qXc3FyFhoa6taenp2vr1q1asWKF1q1bp8rKSs2cOdNe3tjYqOnTp6u+vl65ublasmSJ8vPzlZmZadccPXpU06dPV2xsrN544w1NnjxZCxYsUEFBgV2zefNmZWRkaMaMGcrPz1dYWJhSUlJUU1PT0iEBAACDtCjknDp1So899pgWL16sHj162O0nT55UXl6e5s2bp7i4OEVERCg9PV0lJSUqLS2VJG3fvl3l5eV6+umnFR4eroSEBM2ePVs5OTmqq6uTJOXm5qpfv36aN2+egoKClJycrJEjR+rll1+295WVlaXx48crKSlJwcHBSktLU5cuXZSXl9fy2QAAAMZoUch58sknlZCQoPj4eLf2srIy1dfXu7UHBQWpb9++dsgpLS1VSEiIAgIC7Bqn0ymXy6Xy8nK7Ji4uzm3bTqfT3kZdXZ12797tth8vLy/Fx8erpKSkJUMCAACG6ezpCps2bdI//vEPvf76682WVVdXy8fHR927d3dr9/f3V1VVlV1zYcCRZL++XI3L5dLZs2d14sQJNTY2yt/fv9l+Dh065OmQAACAgTwKOf/617/029/+Vi+99JJuuOGGa9UnAACAq+ZRyNm9e7dqamo0duxYu62xsVEffPCBcnJytHbtWtXX16u2ttbtbE5NTY0CAwMlfXlG5uKnoM4/fXVhzcVPZFVXV8vPz09dunSRl5eXvL29m91kXFNT0+wMEAAA6Jg8uifn9ttv11tvvaWNGzfaPxERERozZoz9bx8fHxUVFdnrHDp0SBUVFYqJiZEkxcTEaP/+/W4BpbCwUH5+fgoODrZriouL3fZdWFhob8PX11cDBw50209TU5OKiorkcDg8mgAAAGAmj87k+Pn5KSQkxK3txhtv1Le+9S27PSkpSUuWLFGPHj3k5+enxYsXy+Fw2AHF6XQqODhYc+bM0WOPPaaqqiqtWLFCkyZNkq+vryRp4sSJysnJ0bJly5SUlKTi4mJt2bJFq1atsvc7depUzZ07VxEREYqKitIrr7yiM2fOuJ1lAgAAHZfHNx5fzvz58+Xl5aXU1FTV1dXJ6XRq4cKF9nJvb2+tXLlSixYt0oQJE9S1a1clJiYqNTXVrunfv79WrVqljIwMZWdnq0+fPlq8eLGGDh1q14waNUrHjh1TZmamqqqqFB4erjVr1nC5CgAASJI6WZZltXUn2orL5dLgwYO1Y8cO+fn5tXV3AFxgdGaBdlfUamDf7tqUOvTyKwDoMK70/ZvvrgIAAEYi5AAAACMRcgAAgJEIOQAAwEiEHAAAYCRCDgAAMBIhBwAAGImQAwAAjETIAQAARiLkAAAAIxFyAACAkQg5AADASIQcAABgJEIOgGumscm6ojYAuBY6t3UHAJjL26uTZueWqLzSJUkK7u2n5yY62rhXADoKQg6Aa6q80qXdFbVt3Q0AHRCXqwAAgJEIOQAAwEiEHAAAYCRCDgAAMBIhB8B1LdDvBh5FB9AiPF0F4LrWvWtnHkUH0CKEHADtAo+iA/AUl6sAAICRCDkAAMBIhBwAAGAkQg4AADASIQcAABiJkAMAAIxEyAEAAEYi5AAAACMRcgAAgJEIOQAAwEiEHAAAYCRCDgAAMBIhBwAAGMmjkPPaa69pzJgxGjRokAYNGqQJEyZo27Zt9vJz584pLS1NsbGxcjgcmjVrlqqrq922UVFRoWnTpik6OlpxcXFaunSpGhoa3Gree+89JSYmKiIiQsOHD9eGDRua9SUnJ0fDhg1TZGSkxo0bp127dnkyFAAAYDiPQk6fPn306KOPasOGDcrLy9Ptt9+uGTNm6MCBA5Kk9PR0bd26VStWrNC6detUWVmpmTNn2us3NjZq+vTpqq+vV25urpYsWaL8/HxlZmbaNUePHtX06dMVGxurN954Q5MnT9aCBQtUUFBg12zevFkZGRmaMWOG8vPzFRYWppSUFNXU1FztfAAAAEN4FHKGDRumhIQE3XLLLRowYIB+/etf68Ybb1RpaalOnjypvLw8zZs3T3FxcYqIiFB6erpKSkpUWloqSdq+fbvKy8v19NNPKzw8XAkJCZo9e7ZycnJUV1cnScrNzVW/fv00b948BQUFKTk5WSNHjtTLL79s9yMrK0vjx49XUlKSgoODlZaWpi5duigvL6/VJgYAALRvLb4np7GxUZs2bdLp06flcDhUVlam+vp6xcfH2zVBQUHq27evHXJKS0sVEhKigIAAu8bpdMrlcqm8vNyuiYuLc9uX0+m0t1FXV6fdu3e77cfLy0vx8fEqKSlp6XAAAIBhOnu6wr59+zRx4kSdO3dON954o1588UUFBwdrz5498vHxUffu3d3q/f39VVVVJUmqrq52CziS7NeXq3G5XDp79qxOnDihxsZG+fv7N9vPoUOHPB0OAAAwlMchZ8CAAdq4caNOnjyp//mf/9HcuXP16quvXou+AQAAtJjHIcfX11f/9m//JkmKiIjQRx99pOzsbP34xz9WfX29amtr3c7m1NTUKDAwUNKXZ2Qufgrq/NNXF9Zc/ERWdXW1/Pz81KVLF3l5ecnb27vZTcY1NTXNzgABAICO66o/J6epqUl1dXWKiIiQj4+PioqK7GWHDh1SRUWFYmJiJEkxMTHav3+/W0ApLCyUn5+fgoOD7Zri4mK3fRQWFtrb8PX11cCBA93209TUpKKiIjkcjqsdDgAAMIRHZ3KeeeYZ/eAHP9B3vvMdnTp1Sn/+85/1/vvva+3aterWrZuSkpK0ZMkS9ejRQ35+flq8eLEcDocdUJxOp4KDgzVnzhw99thjqqqq0ooVKzRp0iT5+vpKkiZOnKicnBwtW7ZMSUlJKi4u1pYtW7Rq1Sq7H1OnTtXcuXMVERGhqKgovfLKKzpz5ozGjh3bejMDAADaNY9CTk1NjebOnavKykp169ZNoaGhWrt2rb7//e9LkubPny8vLy+lpqaqrq5OTqdTCxcutNf39vbWypUrtWjRIk2YMEFdu3ZVYmKiUlNT7Zr+/ftr1apVysjIUHZ2tvr06aPFixdr6NChds2oUaN07NgxZWZmqqqqSuHh4VqzZg2XqwAAgM2jkJOenv61y2+44QYtXLjQLdhc7Lvf/a5Wr179tduJjY3Vxo0bv7YmOTlZycnJX1sDAAA6Lr67CgAAGImQAwAAjETIAQAARiLkAAAAIxFyAACAkQg5AADASIQcAABgJEIOAAAwEiEHAAAYiZADAACMRMgBAABGIuQAAAAjEXIAAICRCDkAAMBIhBwAAGAkQg4AADASIQcAABiJkAMAAIxEyAEAAEYi5AAAACMRcgAAgJEIOQAAwEiEHAAAYCRCDgAAMBIhBwAAGImQAwAAjETIAQAARiLkAAAAIxFyAACAkQg5AADASIQcAN+YQL8b1NhkubVd/BoAWkvntu4AgI6je9fO8vbqpNm5JSqvdCm4t5+em+ho624BMBQhB8A3rrzSpd0VtW3dDQCG43IVAAAwEiEHAAAYiZADAACMRMgBAABG8ijkrFq1SklJSXI4HIqLi9OvfvUrHTp0yK3m3LlzSktLU2xsrBwOh2bNmqXq6mq3moqKCk2bNk3R0dGKi4vT0qVL1dDQ4Fbz3nvvKTExURERERo+fLg2bNjQrD85OTkaNmyYIiMjNW7cOO3atcuT4QAAAIN5FHLef/99TZo0SevXr1dWVpYaGhqUkpKi06dP2zXp6enaunWrVqxYoXXr1qmyslIzZ860lzc2Nmr69Omqr69Xbm6ulixZovz8fGVmZto1R48e1fTp0xUbG6s33nhDkydP1oIFC1RQUGDXbN68WRkZGZoxY4by8/MVFhamlJQU1dTUXM18AAAAQ3gUctauXauxY8fq1ltvVVhYmJYsWaKKigrt3r1bknTy5Enl5eVp3rx5iouLU0REhNLT01VSUqLS0lJJ0vbt21VeXq6nn35a4eHhSkhI0OzZs5WTk6O6ujpJUm5urvr166d58+YpKChIycnJGjlypF5++WW7L1lZWRo/frySkpIUHBystLQ0denSRXl5ea0zMwAAoF27qntyTp48KUnq0aOHJKmsrEz19fWKj4+3a4KCgtS3b1875JSWliokJEQBAQF2jdPplMvlUnl5uV0TFxfnti+n02lvo66uTrt373bbj5eXl+Lj41VSUnI1QwIAAIZocchpampSenq6Bg0apJCQEElSdXW1fHx81L17d7daf39/VVVV2TUXBhxJ9uvL1bhcLp09e1ZffPGFGhsb5e/v32w/F9//AwAAOqYWf+JxWlqaDhw4oNdee601+wMAANAqWnQm58knn9S7776rV155RX369LHbAwICVF9fr9pa949rr6mpUWBgoF1z8dmW868vV+Pn56cuXbqoZ8+e8vb2bnaTcU1NTbMzQAAAoGPyKORYlqUnn3xSf/3rX/XKK6+of//+bssjIiLk4+OjoqIiu+3QoUOqqKhQTEyMJCkmJkb79+93CyiFhYXy8/NTcHCwXVNcXOy27cLCQnsbvr6+GjhwoNt+mpqaVFRUJIeDL/sDAAAehpy0tDS9+eabeuaZZ3TTTTepqqpKVVVVOnv2rCSpW7duSkpK0pIlS1RcXKyysjLNnz9fDofDDihOp1PBwcGaM2eO9u7dq4KCAq1YsUKTJk2Sr6+vJGnixIk6evSoli1bpoMHDyonJ0dbtmzRlClT7L5MnTpV69evV35+vg4ePKhFixbpzJkzGjt2bOvMDAAAaNc8uifnj3/8oyTp/vvvd2vPyMiww8X8+fPl5eWl1NRU1dXVyel0auHChXatt7e3Vq5cqUWLFmnChAnq2rWrEhMTlZqaatf0799fq1atUkZGhrKzs9WnTx8tXrxYQ4cOtWtGjRqlY8eOKTMzU1VVVQoPD9eaNWu4XAUAACR5GHL27dt32ZobbrhBCxcudAs2F/vud7+r1atXf+12YmNjtXHjxq+tSU5OVnJy8mX7BAAAOh6+uwoAABiJkAMAAIxEyAEAAEYi5AAAACMRcgAAgJEIOQAAwEiEHAAAYCRCDgAAMBIhBwAAGImQAwAAjETIAQAARiLkAAAAIxFyAACAkQg5AADASIQcAABgJEIOAAAwEiEHAAAYiZADAACMRMgBAABGIuQAAAAjEXIAAICRCDkAAMBIhBwAAGAkQg4AADASIQcAABiJkAMAAIxEyAEAAEYi5AAAACMRcgAAgJEIOQAAwEiEHAAAYCRCDgAAMBIhBwAAGImQAwAAjETIAQAARiLkAAAAIxFyAACAkTwOOR988IF++ctfyul0KjQ0VG+//bbbcsuy9Nxzz8npdCoqKkpTpkzR4cOH3WqOHz+uRx55RIMGDdKQIUM0f/58nTp1yq1m7969uu+++xQZGamEhAStXr26WV+2bNmiu+66S5GRkRozZoy2bdvm6XAAAIChPA45p0+fVmhoqBYuXPiVy1evXq1169Zp0aJFWr9+vbp27aqUlBSdO3fOrnn00UdVXl6urKwsrVy5Uh9++KGeeOIJe7nL5VJKSor69u2rDRs2aM6cOXrhhRf0pz/9ya75+9//rkceeUT33nuvNm7cqDvuuEMzZszQ/v37PR0SAAAwkMchJyEhQb/+9a81fPjwZsssy1J2drYefPBB3XnnnQoLC9OyZctUWVlpn/E5ePCgCgoKtHjxYkVHR2vIkCFasGCBNm3apM8//1yS9Oabb6q+vl7p6em69dZbNXr0aN1///3Kysqy95Wdna2hQ4fq5z//uYKCgvTQQw/ptttu06uvvtrSuQAAAAZp1XtyPv30U1VVVSk+Pt5u69atm6Kjo1VSUiJJKikpUffu3RUZGWnXxMfHy8vLS7t27ZIklZaWasiQIfL19bVrnE6nPv74Y504ccKuiYuLc9u/0+lUaWlpaw4JAAC0U60acqqqqiRJ/v7+bu3+/v6qrq6WJFVXV6tXr15uyzt37qwePXrY61dXVysgIMCt5vzrC7dzcc2F+wEAAB0bT1cBAAAjtWrICQwMlCTV1NS4tdfU1NhnXQICAnTs2DG35Q0NDTpx4oS9fkBAQLMzMudfX7idi2su3A8AcwX63aDGJsut7eLXANCqIadfv34KDAxUUVGR3eZyubRz5045HA5JksPhUG1trcrKyuya4uJiNTU1KSoqSpIUExOjDz/8UPX19XZNYWGhBgwYoB49etg1xcXFbvsvLCxUTExMaw4JwHWoe9fO8vbqpNm5JRqdWaDZuSXy9urU1t0CcJ3xOOScOnVKe/bs0Z49eyR9ebPxnj17VFFRoU6dOumBBx7Q73//e73zzjvat2+f5syZo969e+vOO++UJAUFBWno0KF6/PHHtWvXLu3YsUNPPfWURo8erW9/+9uSpDFjxsjHx0e/+c1vdODAAW3evFnZ2dmaOnWq3Y8HHnhABQUFeumll3Tw4EE9//zzKisrU3JycmvMC4B2oLzSpd0VtSqvdLV1VwBchzp7ukJZWZkeeOAB+3VGRoYkKTExUUuWLNEvfvELnTlzRk888YRqa2s1ePBgrVmzRjfccIO9zvLly/XUU09p8uTJ8vLy0ogRI7RgwQJ7ebdu3bR27Vo9+eSTGjt2rHr27Klf/epXmjBhgl0zaNAgLV++XCtWrNCzzz6rW265RS+++KJCQkJaNBEAAMAsHoec2NhY7du375LLO3XqpNmzZ2v27NmXrPnWt76lZ5555mv3ExYWptdee+1ra3784x/rxz/+8dd3GMA3prHJ4rIRgOuGxyEHAC7l/H0y5ZUu/TA0UI+NDGvrLgHowHiEHECrOn+fzNFjp9u6KwA6OEIOAAAwEiEHAAAYiZADAACMRMgBAABGIuQAAAAjEXIAAICRCDkAAMBIhBwAAGAkQg4AADASIQcAABiJkAMAAIxEyAEAAEYi5AAAACMRcgAAgJEIOQAAwEiEHAAAYCRCDgAAMBIhBwAAGImQAwAAjETIAQAARiLkAGj3Av1uUGOT1az9q9oAdByd27oDAHC1unftLG+vTpqdW6LySpckKbi3n56b6GjjngFoS4QcAMYor3Rpd0VtW3cDwHWCy1UAAMBIhBwALcL9LgCud1yuAtAiF98D88PQQD02MqyNewUA/4eQA6DFLrwHJijwpjbuDQC443IVAAAwEiEHAAAYiZADAACMRMgBYKSv+hRknggDOhZuPAZwRRqbLHl7dWrrblyxiz8FmU9ABjoeQg6AK3JhYGhPj4vzKchAx8XlKgBX7HxgOHrsdFt3xWN8iSfQ8bT7Mzk5OTlau3atqqqqFBYWpscff1xRUVFt3S0A1xm+xBPoeNr1mZzNmzcrIyNDM2bMUH5+vsLCwpSSkqKampq27hrQrpl8duP82ajdFbU6cbqem5MBg7XrMzlZWVkaP368kpKSJElpaWl69913lZeXp2nTprVx74D24atuKO4oX9nAzcmA2dptyKmrq9Pu3bs1ffp0u83Ly0vx8fEqKSm5om1Y1pd/sblcrmvSR6CtfVWA+aq2le8eVMWJM5KkyH49NG5wfzWcPa2mui/b6s6clsvl0i3dvdRU56Nvd5Xba0nN2q6k5pbuXpf8/6+19/V1NefH6ud1g07UnrzsnF3pvAK4Ns7/3jj/Pn4pnazLVVynPv/8c/3gBz9Qbm6uHI7/+8tr2bJl+uCDD/Tf//3fl93GZ599poSEhGvZTQAAcI1s27ZNffr0ueTydnsmpzX07t1b27Zt00033aROnfgLDACA9sCyLJ06dUq9e/f+2rp2G3J69uwpb2/vZjcZ19TUKCAg4Iq24eXl9bUJEAAAXJ+6det22Zp2+3SVr6+vBg4cqKKiIrutqalJRUVFbpevAABAx9Ruz+RI0tSpUzV37lxFREQoKipKr7zyis6cOaOxY8e2ddcAAEAba9chZ9SoUTp27JgyMzNVVVWl8PBwrVmz5oovVwEAAHO126erAAAAvk67vScHAADg6xByAACAkQg5AADASIQcAABgJELOVTh+/LgeeeQRDRo0SEOGDNH8+fN16tSpS9Z/+umnCg0N/cqfLVu22HVftXzTpk3fxJBazNO5kKT777+/2TifeOIJt5qKigpNmzZN0dHRiouL09KlS9XQ0HAth9IqPJ2P48eP66mnntLIkSMVFRWlH/7wh1q8eLFOnjzpVtdejo2cnBwNGzZMkZGRGjdunHbt2vW19Vu2bNFdd92lyMhIjRkzRtu2bXNbblmWnnvuOTmdTkVFRWnKlCk6fPjwNRxB6/JkPtavX6/77rtP3/ve9/S9731PU6ZMaVY/b968ZsdBSkrKtR5Gq/BkLjZs2NBsnJGRkW41HenY+KrfmaGhoW5fSN2ej41rwkKLpaSkWD/5yU+s0tJS64MPPrCGDx9uPfzww5esb2hosCorK91+nn/+eSsmJsZyuVx2XUhIiJWXl+dWd/bs2W9iSC3m6VxYlmUlJydbCxYscBvnyZMn7eUNDQ3W3XffbU2ZMsX6xz/+Yb377rtWbGys9cwzz1zr4Vw1T+dj37591syZM6133nnHOnLkiFVYWGiNGDHCmjVrlltdezg2Nm3aZA0cONB6/fXXrQMHDlgLFiywhgwZYlVXV39l/Y4dO6zw8HBr9erVVnl5ufVf//Vf1sCBA619+/bZNatWrbIGDx5s/fWvf7X27Nlj/fKXv7SGDRt23Y39q3g6Hw8//LD16quvWv/4xz+s8vJya968edbgwYOtzz77zK6ZO3eulZKS4nYcHD9+/JsaUot5Ohd5eXnWoEGD3MZZVVXlVtORjo0vvvjCbS72799vhYeHW3l5eXZNez02rhVCTguVl5dbISEh1q5du+y2bdu2WaGhoW6/jC7nP//zP63/9//+n1tbSEiI9de//rXV+nqttXQukpOTrcWLF19y+bvvvmuFhYW5/VJ77bXXrEGDBlnnzp1rnc5fA611bGzevNkaOHCgVV9fb7e1h2Pj3nvvtdLS0uzXjY2NltPptFatWvWV9bNnz7amTZvm1jZu3Djr8ccftyzLspqamqzvf//71po1a+zltbW1VkREhPXnP//5GoygdXk6HxdraGiwHA6HlZ+fb7fNnTvXevDBB1u7q9ecp3ORl5dnDR48+JLb6+jHRlZWluVwOKxTp07Zbe312LhWuFzVQiUlJerevbvbqdP4+Hh5eXld9tT8eWVlZdqzZ4/uvffeZsvS0tIUGxure++9V6+//vplv06+LV3NXLz11luKjY3V3XffrWeeeUZnzpyxl5WWliokJMTtwx2dTqdcLpfKy8tbfyCtpDWODUlyuVzy8/NT587un9l5PR8bdXV12r17t+Lj4+02Ly8vxcfHq6Sk5CvXKS0tVVxcnFub0+lUaWmppC8v81ZVVblts1u3boqOjr7kNq8XLZmPi505c0YNDQ3q0aOHW/v777+vuLg4jRw5UgsXLtQXX3zRqn1vbS2di9OnT+tHP/qREhIS9OCDD+rAgQP2so5+bOTl5Wn06NG68cYb3drb27FxLbXrTzxuS9XV1erVq5dbW+fOndWjRw9VVVVd0TZef/11BQUFadCgQW7tqampuv3229W1a1dt375daWlpOn36tB544IFW639raulc3H333erbt6969+6tffv2afny5fr444/1wgsv2Nu9+NOrz7++0jluC61xbBw7dky/+93vNGHCBLf26/3Y+OKLL9TY2Ch/f3+3dn9/fx06dOgr1/mq/87+/v6qrq6W9H//rb9qm+drrlctmY+LLV++XL1793Z7Mxw6dKiGDx+ufv366ejRo3r22Wf1i1/8Qn/605/k7e3dqmNoLS2ZiwEDBig9PV2hoaE6efKkXnrpJU2cOFGbNm1Snz59OvSxsWvXLu3fv1+//e1v3drb47FxLRFyLrJ8+XKtXr36a2s2b9581fs5e/as/vznP+tXv/pVs2UzZsyw/33bbbfpzJkzWrt27Tf+Rnat5+LCN/DQ0FAFBgZqypQp+uSTT3TzzTe3eLvXyjd1bLhcLk2fPl1BQUGaOXOm27Lr5djAN+MPf/iDNm/erOzsbN1www12++jRo+1/n7+59M4777T/gjeFw+Fw+8Jlh8OhUaNGKTc3Vw899FDbdew68PrrryskJERRUVFu7R3l2LhShJyL/OxnP1NiYuLX1vTv318BAQE6duyYW3tDQ4NOnDihwMDAy+7nL3/5i86ePat77rnnsrXR0dH63e9+p7q6Ovn6+l62vrV8U3NxXnR0tCTpyJEjuvnmmxUQENDs8s75v8482W5r+Sbmw+Vy6ec//7luuukmvfjii/Lx8fna+rY6Ni6lZ8+e8vb2Vk1NjVt7TU3NJb9TLiAgoNlf3RfWn5+zmpoa9e7d260mLCysNbvf6loyH+etXbtWf/jDH5SVlXXZcfbv3189e/bUkSNHrts3squZi/N8fHwUHh6uTz75RFLHPTZOnz6tTZs2KTU19bL7aQ/HxrXEPTkX6dWrl4KCgr72x9fXVw6HQ7W1tSorK7PXLS4uVlNTU7Nk/VXy8vI0bNiwZpc1vsqePXvUo0ePb/xN7Juai/P27Nkj6f9+ccXExGj//v1uvwQKCwvl5+en4ODgVhrllbvW8+FyuZSSkiIfHx/9/ve/d/vL/VLa6ti4FF9fXw0cOFBFRUV2W1NTk4qKitz+Ir9QTEyMiouL3doKCwsVExMjSerXr58CAwPdtulyubRz585LbvN60ZL5kKTVq1frd7/7ndasWdPskemv8tlnn+n48eNtEv6vVEvn4kKNjY3av3+/Pc6OeGxIX/6RXFdXp5/85CeX3U97ODauqba+87k9S0lJse655x5r586d1ocffmiNGDHC7THhzz77zBo5cqS1c+dOt/UOHz5shYaGWtu2bWu2zXfeecdav369tW/fPuvw4cNWTk6OFR0dbT333HPXfDxXw9O5OHLkiPXCCy9YH330kXX06FHr7bfftu644w5r0qRJ9jrnHyH/2c9+Zu3Zs8f63//9X+v2229vN4+QezIfJ0+etMaNG2fdfffd1pEjR9we/2xoaLAsq/0cG5s2bbIiIiKsDRs2WOXl5dbjjz9uDRkyxH5K7rHHHrOWL19u1+/YscO67bbbrLVr11rl5eVWZmbmVz5CPmTIEOvtt9+29u7daz344IPt6jFhT+Zj1apV1sCBA62//OUvbsfB+Y+ZcLlc1pIlS6ySkhLr6NGjVmFhoZWYmGiNGDHiun7q0LI8n4vnn3/eKigosD755BOrrKzM+vWvf21FRkZaBw4csGs60rFx3k9/+lProYceatbeno+Na4XLVVdh+fLleuqppzR58mR5eXlpxIgRWrBggb28vr5eH3/8sdsTQ9KXZ3H69Okjp9PZbJudO3dWTk6O0tPTJUk333yz5s2bp/Hjx1/bwVwlT+fCx8dHRUVFys7O1unTp/Wd73xHI0aMcLtHydvbWytXrtSiRYs0YcIEde3aVYmJiVd0irateTofu3fv1s6dOyVJw4cPd9vWO++8o379+rWbY2PUqFE6duyYMjMzVVVVpfDwcK1Zs8Y+Bf+vf/1LXl7/dxJ50KBBWr58uVasWKFnn31Wt9xyi1588UWFhITYNb/4xS905swZPfHEE6qtrdXgwYO1Zs2aKzrb1dY8nY/c3FzV19c3O85nzpypWbNmydvbW/v379fGjRt18uRJ9e7dW9///vc1e/bs6+aM3qV4Ohe1tbV6/PHHVVVVpR49emjgwIHKzc11O5PbkY4NSTp06JB27Nihl156qdn22vOxca10sqzr6PlTAACAVsI9OQAAwEiEHAAAYCRCDgAAMBIhBwAAGImQAwAAjETIAQAARiLkAAAAIxFyAACAkQg5AADASIQcAABgJEIOAAAwEiEHAAAY6f8DDecMneF6LlIAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "aa = plt.hist(co_grn.weight, bins=100)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "co_grn.to_csv(f'{work_dir}/infer/celloracle/grn/co_grn{tag}.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "celloracle",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
